streamer/pipe.py

```diff
@@ -0,0 +1,146 @@
+from threading import Thread
+from typing import Any
+
+class Pipe(Thread):
+  """A class that represents a pipe."""
+
+  def __init__(self):
+    """Initializes a non-functioning pipe."""
+    
+    self._read_pipe_name = ''
+    self._write_pipe_name = ''
+
+    # Windows specific declarations.
+    self.buf_size = 0
+    self._read_side: Any = None
+    self._write_side: Any = None
+
+  @staticmethod
+  def create_ipc_pipe(temp_dir: str, suffix: str = '') -> 'Pipe':
+    """A static method used to create a pipe between two processes. 
+    
+    On POXIS systems, it creates a named pipe using `os.mkfifo`.
+    
+    On Windows platforms, it starts a backgroud thread that transfars data from the
+    writer to the reader process it is connected to.
+    """
+
+    unique_name = str(uuid.uuid4()) + suffix
+    pipe = Pipe()
+
+    # New Technology, aka WindowsNT.
+    if os.name == 'nt':
+      import win32pipe # type: ignore
+      # Initialize the pipe object as a thread with `daemon` attribute set
+      # to True so that the thread shuts down when the caller thread exits.
+      Thread.__init__(pipe, daemon=True)
+      pipe_name = '-nt-shaka-' + unique_name
+      # The read pipe is connected to a writer process.
+      pipe._read_pipe_name = r'\\.\pipe\W' + pipe_name
+      # The write pipe is connected to a reader process.
+      pipe._write_pipe_name = r'\\.\pipe\R' + pipe_name
+      pipe.buf_size = 64 * 1024
+
+      pipe._read_side = win32pipe.CreateNamedPipe(
+          pipe._read_pipe_name,
+          win32pipe.PIPE_ACCESS_INBOUND,
+          win32pipe.PIPE_WAIT | win32pipe.PIPE_TYPE_BYTE | win32pipe.PIPE_READMODE_BYTE,
+          1,
+          pipe.buf_size,
+          pipe.buf_size,
+          0,
+          None)
+
+      pipe._write_side = win32pipe.CreateNamedPipe(
+          pipe._write_pipe_name,
+          win32pipe.PIPE_ACCESS_OUTBOUND,
+          win32pipe.PIPE_WAIT | win32pipe.PIPE_TYPE_BYTE | win32pipe.PIPE_READMODE_BYTE,
+          1,
+          pipe.buf_size,
+          pipe.buf_size,
+          0,
+          None)
+
+      # Start the thread.
+      pipe.start()
+    elif hasattr(os, 'mkfifo'):
+      pipe_name = os.path.join(temp_dir, unique_name)
+      pipe._read_pipe_name = pipe_name
+      pipe._write_pipe_name = pipe_name
+      readable_by_owner_only = 0o600  # Unix permission bits
+      os.mkfifo(pipe_name, mode=readable_by_owner_only) # type: ignore
+    else:
+      raise RuntimeError('Platform not supported.')
+    return pipe
+
+  @staticmethod
+  def create_file_pipe(path: str, mode: str) -> 'Pipe':
+    """Returns a Pipe object whose read or write end is a path to a file."""
+
+    pipe = Pipe()
+    # A process will write on the read pipe(file).
+    if mode == 'w':
+      pipe._read_pipe_name = path
+    # A process will read from the write pipe(file).
+    elif mode == 'r':
+      pipe._write_pipe_name = path
+    else:
+      raise RuntimeError('{} is not a valid file mode'.format(mode))
```

Comment:
I think it's unlikely, but in case someone passes a mode like 'rw' or 'wb' or 'a+', which are all valid modes for real files, let's change this to say something like "not a valid mode for Pipe".